csharp/src/Apache.Arrow/ArrowBuffer.BitPackedBuilder.cs

```diff
@@ -0,0 +1,268 @@
+            {
+                CheckIndex(index);
+                BitUtility.SetBit(Span, index);
+                return this;
+            }
+
+            /// <summary>
+            /// Set the bit at a particular index to a given value.
+            /// </summary>
+            /// <param name="index">Index of bit to set/unset.</param>
+            /// <param name="value">Value of bit.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
+            public BitPackedBuilder Set(int index, bool value)
+            {
+                CheckIndex(index);
+                BitUtility.SetBit(Span, index, value);
+                return this;
+            }
+
+            /// <summary>
+            /// Swap the bits at two given indices.
+            /// </summary>
+            /// <param name="i">First index.</param>
+            /// <param name="j">Second index.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
+            public BitPackedBuilder Swap(int i, int j)
+            {
+                CheckIndex(i);
+                CheckIndex(j);
+                var bi = BitUtility.GetBit(Span, i);
+                var bj = BitUtility.GetBit(Span, j);
+                BitUtility.SetBit(Span, i, bj);
+                BitUtility.SetBit(Span, j, bi);
+                return this;
+            }
+
+            /// <summary>
+            /// Reserve a given number of bits' additional capacity.
+            /// </summary>
+            /// <param name="bitAdditionalCapacity">Number of bits of required additional capacity.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
+            public BitPackedBuilder Reserve(int bitAdditionalCapacity)
+            {
+                if (bitAdditionalCapacity < 0)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(bitAdditionalCapacity));
+                }
+
+                EnsureAdditionalCapacity(bitAdditionalCapacity);
+                return this;
+            }
+
+            /// <summary>
+            /// Resize the buffer to a given size.
+            /// </summary>
+            /// <remarks>
+            /// Note that if the required capacity is smaller than the current length of the populated buffer so far,
+            /// the buffer will be truncated and items at the end of the buffer will be lost.
+            /// </remarks>
+            /// <remarks>
+            /// Note also that a negative capacity will result in the buffer being resized to zero.
+            /// </remarks>
+            /// <param name="bitCapacity">Number of bits of required capacity.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
+            public BitPackedBuilder Resize(int bitCapacity)
+            {
+                bitCapacity = bitCapacity < 0 ? 0 : bitCapacity;
```

Comment:
Replicating existing behaviour for negative capacities, as seen on `ArrowBuffer.Builder`.

See other comment on `ArrowBuffer.Builder.Resize()`: do we need to replicate this behaviour?